socs/db/suprsync.py

```diff
@@ -0,0 +1,355 @@
+        if res.stderr:
+            self.log.error("stderr for cmd: {cmd}\n{err}",
+                           cmd=_cmd, err=res.stderr.decode())
+        # To many ssh commands in to short a period will break things
+        time.sleep(1)
+        return res
+
+    def copy_file(self, file):
+        """
+        Attempts to copy a file to its dest.
+
+        Args
+        ----
+        file : SupRsyncFile
+            file to be copied
+
+        Returns
+        --------
+        res : str or bool
+            Returns False if unsuccessful, and the md5sum calculated on the
+            remote server if successful.
+        """
+        remote_path = os.path.join(self.remote_basedir, file.remote_path)
+
+        # Creates directory on dest server:
+        res = self.run_on_remote(['mkdir', '-p', os.path.dirname(remote_path)])
+        if res.returncode != 0:
+            self.log.error("remote mkdir failed:\n{e}", e=res.stderr)
+            return False
+
+        if self.ssh_host is not None:
+            dest = self.ssh_host + ':' + remote_path
+        else:
+            dest = remote_path
+
+        cmd = ['rsync', '-t']
+        if self.ssh_key is not None:
+            cmd.extend(['--rsh', f'ssh -i {self.ssh_key}'])
+
+        cmd.extend([file.local_path, dest])
+        self.log.debug("Running: " + ' '.join(cmd))
+        res = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                             timeout=self.copy_timeout)
+        time.sleep(1)
+
+
+        if res.returncode != 0:
+            self.log.error("Rsync failed for file {path}!\nstderr: {err}",
+                           path=file.local_path, err=res.stderr.decode())
+            return False
+
+        res = self.run_on_remote(['md5sum', remote_path])
+        if res.returncode != 0:
+            return False
+        else:
+            return res.stdout.decode().split()[0]
+
+    def handle_file(self, file, session):
+        """
+        Handles operation of an un-removed SupRsyncFile. Will attempt to copy,
+        calculate the remote md5sum, and remove from the local host if enough
+        time has passed.
+
+        Args
+        ----
+        file : SupRsyncFile
+            file to be copied
+        """
+        # File must have been removed for some reason
+        if not os.path.exists(file.local_path):
+            self.log.warn(
+                "File {path} does not exist! Setting removed time to 0.",
+                path=file.local_path
+            )
+            file.removed = 0
+            return
+
+        # Check that file md5sum matches whatever it was before
+        current_md5sum = get_md5sum(file.local_path)
+        if current_md5sum != file.local_md5sum:
+            self.log.warn(
+                "Md5sum of {path} has changed! Resetting it in the db and "
+                "trying to copy", path=file.local_path)
+            file.local_md5sum = current_md5sum
```

Comment:
Files are usually only registered after they finished writing. This mainly happens when a file is written and registered, and then written over or updated before being archived. This definitely isn't ideal, but has happened before and causes problems since if the local file changes before being archived, the agent will never be able to rsync successfully since it's comparing against a different checksum.